### PR TITLE
Improved `IN PARTITION` mutations

### PR DESCRIPTION
**Pull request by quickhouse**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improved `IN PARTITION` mutations.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/StorageMergeTree.cpp [hunk 1]
```diff
@@ -431,6 +431,14 @@ CurrentlyMergingPartsTagger::~CurrentlyMergingPartsTagger()
 
 Int64 StorageMergeTree::startMutation(const MutationCommands & commands, ContextPtr query_context)
 {
+    /// Validate partition IDs (if any) before starting mutation.
+    const auto & set = getPartitionIdsAffectedByCommands(commands, query_context);
+    std::optional<String> partition_id;
+    if (set.size() == 1)
+    {
+        partition_id.emplace(std::move(*set.begin()));
+    }
```
Comment (filimonov) [inline src/Storages/StorageMergeTree.cpp:440]: else?

File: src/Storages/StorageMergeTree.h [hunk 2]
```diff
@@ -145,7 +145,198 @@ class StorageMergeTree final : public MergeTreeData
     /// This set have to be used with `currently_processing_in_background_mutex`.
     DataParts currently_merging_mutating_parts;
 
-    std::map<UInt64, MergeTreeMutationEntry> current_mutations_by_version;
+    class CurrentMutations
+    {
+        using MutationByVersionMap = std::map<UInt64, MergeTreeMutationEntry>;
+        using MutationReferenceByVersionMap = std::map<UInt64, std::reference_wrapper<MergeTreeMutationEntry>>;
+        using MutationReferenceByPartitionByVersionMap = std::unordered_map<String, MutationReferenceByVersionMap>;
```
Comment (filimonov) [inline src/Storages/StorageMergeTree.h:152]: some comments would be nice

File: src/Storages/StorageMergeTree.h [hunk 3]
```diff
@@ -145,7 +145,198 @@ class StorageMergeTree final : public MergeTreeData
...
+
+        MutationByVersionMap all_mutations;
+        MutationReferenceByPartitionByVersionMap mutations_by_partition;
```
Comment (filimonov) [inline src/Storages/StorageMergeTree.h:155]: I think it would be better to build it dynamically (by filtering out from all_mutations items which are not related to the partition). 

Why: less RAM needed (no need to duplicate mutations for every partition), single source of truth

File: src/Storages/StorageMergeTree.h [hunk 4]
```diff
@@ -145,7 +145,198 @@ class StorageMergeTree final : public MergeTreeData
...
+            else
+            {
+                return empty_mutation_map;
+            }
+        }
+
+        /// Uniqueness of versions comes from caller code.
+        auto try_emplace(Int64 version, MergeTreeMutationEntry && entry)
+        {
+            auto result = all_mutations.try_emplace(version, std::move(entry));
```
Comment (filimonov) [inline src/Storages/StorageMergeTree.h:211]: thread safety?

File: src/Storages/StorageMergeTree.h [hunk 5]
```diff
@@ -145,7 +145,198 @@ class StorageMergeTree final : public MergeTreeData
...
+                    for (auto & [_, mutations] : mutations_by_partition)
+                    {
+                        mutations.emplace(version, result.first->second);
+                    }
+                }
+            }
+            return result;
+        }
+
+        // FIXME Как появляются новые партишны в мапе??
```
Comment (filimonov) [inline src/Storages/StorageMergeTree.h:229]: fix :)

File: src/Storages/StorageMergeTree.h [hunk 6]
```diff
@@ -145,7 +145,198 @@ class StorageMergeTree final : public MergeTreeData
...
+        }
+
+        auto lower_bound(UInt64 version)
+        {
+            return all_mutations.lower_bound(version);
+        }
+
+        auto upper_bound(UInt64 version)
+        {
+            return all_mutations.upper_bound(version);
```
Comment (filimonov) [inline src/Storages/StorageMergeTree.h:329]: hm... shouldn't it take one more arg (partition_id)

File: src/Storages/StorageMergeTree.h (hunk 4 above)
Comment (quickhouse) [reply src/Storages/StorageMergeTree.h]: It's safe. :)

File: src/Storages/StorageMergeTree.h (hunk 6 above)
Comment (quickhouse) [reply src/Storages/StorageMergeTree.h]: This one was for all partitions.

File: src/Storages/StorageMergeTree.cpp (hunk 1 above)
Comment (filimonov) [reply src/Storages/StorageMergeTree.cpp:447]: So for example 3 is not possible?

If it can return empty or single element list - why it return list? May be it should use optional?

File: src/Storages/StorageMergeTree.cpp [hunk 7]
```diff
@@ -1831,9 +1855,16 @@ MutationCommands StorageMergeTree::getFirstAlterMutationCommandsForPart(const Da
     std::lock_guard lock(currently_processing_in_background_mutex);
 
     auto it = current_mutations_by_version.upper_bound(part->info.getDataVersion());
-    if (it == current_mutations_by_version.end())
-        return {};
-    return it->second.commands;
+    while (it != current_mutations_by_version.end())
```
Comment (filimonov) [inline src/Storages/StorageMergeTree.cpp:1858]: Not sure if it's the problem, but it seem like in the pessimistic scenario (100000 parts of a version lower than max), that `while` will be executed quite often...

Comment (tavplubix) [inline src/Storages/StorageMergeTree.cpp:1858]: Why do we need to change this function at all? `getFirstAlterMutationCommandsForPart` is something too specific and seems like it used only for ALTER METADATA, not for ALTER UPDATE/DELETE. Is it possible to ADD COLUMN IN PARTITION, for example? (seems like it's not). I do not understand why these changes are related to ALTER UPDATE/DELETE IN PARTITION.

**Comment by tavplubix** [comment]
Seems like this implementation is something completely different from [the solution](https://gist.github.com/filimonov/a2c79385159efe2935e616ac7492632b) proposed by @filimonov. It it expected? Please add a comment with something like design doc describing you solution and explaining how (and why) it works.

File: src/Storages/MergeTree/MergeTreeMutationEntry.cpp [hunk 8]
```diff
@@ -45,21 +48,35 @@ UInt64 MergeTreeMutationEntry::parseFileName(const String & file_name_)
...
+    const TransactionID & tid_,
+    const WriteSettings & settings
+)
     : create_time(time(nullptr))
     , commands(std::move(commands_))
     , disk(std::move(disk_))
     , path_prefix(path_prefix_)
     , file_name("tmp_mutation_" + toString(tmp_number) + ".txt")
     , is_temp(true)
+    , partition_id(partition_id_)
```
[the remark this answers is not in the document]
Comment (quickhouse) [reply src/Storages/MergeTree/MergeTreeMutationEntry.cpp:66]: We can, but that function is already called in the only caller of this method (`startMutation()`).

**Comment by quickhouse** [comment]
> Seems like this implementation is something completely different from [the solution](https://gist.github.com/filimonov/a2c79385159efe2935e616ac7492632b) proposed by @filimonov. Is it expected? Please add a comment with something like design doc describing you solution and explaining how (and why) it works.
I fixed this in 47c9fd519c0604371f3b18577236151df18aae3d . Map by partition allows to stop iteration by `current_mutations_by_version` early.

File: src/Storages/StorageMergeTree.cpp (hunk 7 above)
Comment (quickhouse) [reply src/Storages/StorageMergeTree.cpp:1886]: This function (unexpectedly) returns next (upcoming) mutation for specified part. I would leave it as in branch for consistency at least. Otherwise `getFirstAlterMutationCommandsForPart()` would return irrelevant mutations (for different partitions).

Comment (quickhouse) [reply src/Storages/StorageMergeTree.cpp:1886]: Fixed in 47c9fd519c0604371f3b18577236151df18aae3d .

File: src/Storages/StorageMergeTree.cpp (hunk 1 above)
Comment (quickhouse) [reply src/Storages/StorageMergeTree.cpp:440]: This is a common code with `StorageReplicatedMergeTree`, we can't change it.

Comment (tavplubix) [reply src/Storages/StorageMergeTree.cpp:440]: Nobody asks to change it. Please add a comment that explains why this function always returns one partition for `StorageMergeTree` (I don't understand why). And throw LOGICAL_ERROR if it returned more than one partition.

File: src/Storages/MergeTree/MergeTreeMutationEntry.cpp [hunk 9]
```diff
@@ -45,21 +48,35 @@ UInt64 MergeTreeMutationEntry::parseFileName(const String & file_name_)
...
     {
         auto out = disk->writeFile(std::filesystem::path(path_prefix) / file_name, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Rewrite, settings);
         *out << "format version: 1\n"
             << "create time: " << LocalDateTime(create_time) << "\n";
+        if (partition_id)
+        {
+            *out << "partition id: ";
+            writeEscapedString(*partition_id, *out);
+            *out << "\n";
+        }
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreeMutationEntry.cpp:79]: Do we really need to serialize `partition_id` here? Seems like we already have partition id inside `commands`

File: src/Storages/StorageMergeTree.h [hunk 10]
```diff
@@ -145,6 +145,14 @@ class StorageMergeTree final : public MergeTreeData
     /// This set have to be used with `currently_processing_in_background_mutex`.
     DataParts currently_merging_mutating_parts;
 
+    /// Accessed under `currently_processing_in_background_mutex` lock.
```
Comment (tavplubix) [inline src/Storages/StorageMergeTree.h:148]: Consider using [TSA annotations](https://github.com/ClickHouse/ClickHouse/blob/cedf75ed5ebf2912490200ee82b6a110afc7bb70/base/base/defines.h#L137-L147)

File: src/Storages/StorageMergeTree.cpp [hunk 11]
```diff
@@ -457,6 +465,17 @@ Int64 StorageMergeTree::startMutation(const MutationCommands & commands, Context
         if (!inserted)
             throw Exception(ErrorCodes::LOGICAL_ERROR, "Mutation {} already exists, it's a bug", version);
 
+        if (partition)
+        {
+            last_mutation_by_partition[*partition] = version;
```
Comment (tavplubix) [inline src/Storages/StorageMergeTree.cpp:470]: Do we need to update this map on server startup as well? Please add an integration tests that runs mutation in partition and restarts server.

Comment (quickhouse) [reply src/Storages/StorageMergeTree.cpp:470]: Yes, it can improve performance in case there are lots of unfinished mutations for specific partitions, but strictly is not necessary. I'll make it in separate commit.

Comment (quickhouse) [reply src/Storages/StorageMergeTree.cpp:470]: bcfc588ed07e36e530b3f47e8d50e261cbd3f07b

File: src/Storages/MergeTree/MergeTreeMutationEntry.cpp (hunk 9 above)
Comment (quickhouse) [reply src/Storages/MergeTree/MergeTreeMutationEntry.cpp:78]: We don't need serialization of it specifically, but we need a way to distinguish old storage-wide mutations from new. If we don't change format, we will try to recalculate `paritition_id` for older and possibly unfinished partition-specific mutations, skip unrelated partitions and stick on merges when in the same partition subset of parts is mutated beyond correct version, or stick being unable to mutate another subset beyond it because there is no related mutation for that. I think this serialization is a neat way to avoid that.

File: src/Storages/StorageMergeTree.h (hunk 10 above)
Comment (quickhouse) [reply src/Storages/StorageMergeTree.h:148]: Wow, cool!

Comment (quickhouse) [reply src/Storages/StorageMergeTree.h:148]: @tavplubix it does not quite fit in this case because there is (at least) one usage of `unique_lock` instead of `lock_guard` in `selectPartsToMerge()` and we would not be able to build complete tree of function calls properly annotated.

https://lists.llvm.org/pipermail/cfe-dev/2016-November/051468.html

File: src/Storages/StorageMergeTree.cpp (hunk 1 above)
Comment (tavplubix) [reply src/Storages/StorageMergeTree.cpp:440]: Please add a comment

File: src/Storages/StorageMergeTree.h (hunk 10 above)
Comment (tavplubix) [reply src/Storages/StorageMergeTree.h:148]: You can use `TSA_SUPPRESS_WARNING_FOR_READ` in `selectPartsToMerge`

File: src/Storages/StorageMergeTree.cpp (hunk 11 above)
Comment (tavplubix) [reply src/Storages/StorageMergeTree.cpp:470]: > Yes, it can improve performance in case there are lots of unfinished mutations for specific partitions, but strictly is not necessary

You're joking, right? According to the comment above you've implemented `partition_id` serialization in order to distinguish old and new IN PARTITION mutations after server restart. But it does not make any sense without actually using these `partition_id`s on startup.

**Comment by quickhouse** [comment]
Problem with exceptions was caused by lack of error handling in Kazoo ZooKeeper client python-zk/kazoo#672 . By default `use_keeper=True` for all test instances. Example: https://gist.github.com/quickhouse/09d6da34f35a13024b1f3432c6969c03

**Comment by quickhouse** [comment]
Problem with exceptions is caused by wrong interoperability of `logging` and `pytest` and seems unfixable pytest-dev/pytest#5502 . I would recommend to throw `pytest` away. The problem may randomly appear in any test in test suite. Example of output: https://gist.github.com/quickhouse/09d6da34f35a13024b1f3432c6969c03 .

File: src/Storages/StorageMergeTree.cpp (hunk 11 above)
Comment (quickhouse) [reply src/Storages/StorageMergeTree.cpp:470]: No, actually I implemented `partition_id` in order to have more straightforward approach and not calculate it every load (but only one time for mutation). But, it helped in distinguishing old and new mutations as well.

I'm not joking here, it actually not necessary for correctness, but improves performance in some use cases. This is so because if something is not found in map then search algorithm will simply go until the end of container (or `upper_bound()`, checking if `partition_id` is suitable. Basically, this map only saves iterations in case when there are many mutations for specific partitions, but does not affect correctness.

File: src/Storages/StorageMergeTree.h (hunk 10 above)
Comment (quickhouse) [reply src/Storages/StorageMergeTree.h:148]: This is impossible because `std::unique_lock` has no thread annotations. There is nothing to suppress, however, but it is needed to mark mutex as locked for Clang instead. It would work if we only used `std::lock_guard`.

File: src/Storages/MergeTree/MergeTreeMutationEntry.cpp (hunk 9 above)
Comment (quickhouse) [reply src/Storages/MergeTree/MergeTreeMutationEntry.cpp:78]: I mean that we have different mutations:
1) with `partition_id` and `IN PARTITION` expression, which are new partition-specific mutations — these shall create partition-specific changes;
2) without `partition_id` and `IN PARTITION` expression, which are old or new storage-wide mutations — these create storage-wide changes;
3) without `partition_id` but with `IN PARTITION` expression, which are old partition-specific mutations. They were supposed to make storage-wide changes earlier (before this patch) and may be partially finished on some parts, so we must continue working the same way on them. If we remove these guys from remaining really unaffected parts (`parts_to_do`), we would have to rollback unaffected parts which already mutated to that version, or we will have finished mutation and different versions in a partition which will prevent merges to happen until new mutation (in that partition or whole storage) appears.

Generally, I think this is a good thing to have partition set to be calculated at the moment when the mutation is created.

File: src/Storages/StorageMergeTree.h (hunk 10 above)
Comment (tavplubix) [reply src/Storages/StorageMergeTree.h:148]: It is possible and it works fine with `std::unique_lock` and `TSA_SUPPRESS_WARNING_FOR_READ`

File: src/Storages/MergeTree/MergeTreeMutationEntry.cpp (hunk 9 above)
Comment (tavplubix) [reply src/Storages/MergeTree/MergeTreeMutationEntry.cpp:78]: I understand your point, but I'm not sure about

> we must continue working the same way on them. If we remove these guys from remaining really unaffected parts (parts_to_do), we would have to rollback unaffected parts which already mutated to that version

I agree that it's undesirable to have some parts with data version greater than expected version for the partition. But AFAIK it will not break anything. 

> will prevent merges to happen until new mutation (in that partition or whole storage) appears

It will prevent only merges of parts with different data versions, seems like it's not critical. Also it probably can be addressed by taking `last_mutation_by_partition` into account in `getCurrentMutationVersion`.

I don't like serializing partition id for two reasons:
 - It duplicates partition id that we already have in `IN PARTITION` clause (and we already serialize it).
 - It unconditionally changes format of mutation entries without any settings. It will make downgrade impossible once you run `IN PARTITION` mutation. `IN PARTITION` is not a new feature, so this change is backward incompatible. We can introduce a setting like `merge_tree_mutation_entry_version` or `serialize_partition_id_for_in_partition_mutations` for compatibility, but I don't like this option. It's better to avoid changing format, and seems like it's possible to avoid it.

**Comment by alexey-milovidov** [comment]
> Improved IN PARTITION mutations.

@quickhouse how exactly?

**Comment by quickhouse** [comment]
I apologize, I am completely exhausted and can't look at this in the near
future.

**Comment by filimonov** [comment]
@alexey-milovidov the implementation of ALTER IN PARTITION for MergeTree was incorrect and was rolled back https://github.com/ClickHouse/ClickHouse/pull/40589 

That PR was implementing that concept https://gist.github.com/filimonov/a2c79385159efe2935e616ac7492632b (discussed with @tavplubix )

But since Vladimir ( @quickhouse ) did not finish that task, I'm closing the PR we will reassign the task to finish it.

**Comment by aadant** [comment]
@filimonov : please also fix the user experience of this feature, if the IN PARTITION is touching selected parts (by virtue of partition pruning), Current implementation would show all parts (I wonder if it is even processing them as I see the counter decrease over time !). My test was mutating 1 partition and I saw thousands of them in system.mutations.
